### PR TITLE
Add multiple queries and checks to generic query measurement

### PR DESCRIPTION
**Pull request by aleksandra-malinowska**
Follow-up to #1957. Changes:
- metric is no longer treated as latency,
- multiple queries can be defined for the same measurement instance,
- user can set threshold - all samples are expected to be lesser or equal to it.

@jprzychodzen - can you PTAL and see if the basic metric summary (avg, min, max) is what you meant? I'm not entirely sure what is the use case here (it obviously doesn't do anything for queries with top-level histogram_quantile()). If we're mostly supporting queries returning a single sample, perhaps we could instead return DataItem with names of queries as keys and results as values?

### CODE REVIEW

File: clusterloader2/pkg/measurement/common/generic_query_measurement.go [hunk 1]
```diff
@@ -68,39 +104,58 @@ func (g *genericQueryGatherer) IsEnabled(config *measurement.Config) bool {
...
-	metric, err := g.query(executor, startTime, endTime)
+	var errs []error
+	data := map[string]float64{}
+	for _, q := range g.queries {
+		samples, err := g.query(q, executor, startTime, endTime)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(samples) != 1 {
```
Comment (mborsz) [inline clusterloader2/pkg/measurement/common/generic_query_measurement.go:115]: I'm wondering if also len(samples) == 0 should be accepted.

Let's take a case like I want to track latency of POST on pods, so I write query like:
```
        - name: Perc99
          query: histogram_quantile(0.99, sum(rate(apiserver_request_duration_seconds_bucket{verb="POST", resource="pods}[%v])) by (le))
          threshold: 60
```

Let's say that there are no post on pods. I would expect this to not fail (as there is no sample violating the theshold).

For summary/perfdash you can log any value you want (e.g. not create a key for a query with 0 samples or log 0.0 or NaN).

File: clusterloader2/pkg/measurement/common/generic_query_measurement.go [hunk 2]
```diff
@@ -40,26 +45,57 @@ func init() {
...
+	for _, q := range queries {
+		query, err := util.GetString(q, "query")
+		if err != nil {
+			return err
+		}
+		name, err := util.GetString(q, "name")
+		if err != nil {
+			return err
+		}
+		threshold, err := util.GetFloat64(q, "threshold")
```
Comment (mborsz) [inline clusterloader2/pkg/measurement/common/generic_query_measurement.go:88]: Does it mean that theshold is a required field (I think util.GetFloat64 returns an error otherwise).

I think we should allow metric not to have any thresholds (like e.g. simply tracking some value, without validation)?

File: clusterloader2/pkg/measurement/common/generic_query_measurement.go [hunk 3]
```diff
@@ -68,39 +104,58 @@ func (g *genericQueryGatherer) IsEnabled(config *measurement.Config) bool {
 }
 
 func (g *genericQueryGatherer) Gather(executor QueryExecutor, startTime, endTime time.Time, config *measurement.Config) ([]measurement.Summary, error) {
```
Comment (mborsz) [inline clusterloader2/pkg/measurement/common/generic_query_measurement.go:106]: Can we add some unit tests for this method? They can simply use fakeQueryExecutor to returns some artificial samples. The goal is to validate:
* that threshold is properly validated (no threshold, threshold higher than value, threshold lower than value)
* that the conversion to the summary happens as we want
* what happens if there are len(samples) > 1
* what happens when len(samples) == 0

File: clusterloader2/pkg/measurement/common/generic_query_measurement.go (hunk 1 above)
Comment (mborsz) [reply clusterloader2/pkg/measurement/common/generic_query_measurement.go:115]: not creating a key for a query in case of 0 samples seems to be the cleanest solution IMHO

File: clusterloader2/pkg/measurement/common/generic_query_measurement.go (hunk 2 above)
Comment (aleksandra-malinowska) [reply clusterloader2/pkg/measurement/common/generic_query_measurement.go:90]: Done

File: clusterloader2/pkg/measurement/common/generic_query_measurement.go (hunk 1 above)
Comment (aleksandra-malinowska) [reply clusterloader2/pkg/measurement/common/generic_query_measurement.go]: Added skipping queries that return no results

File: clusterloader2/pkg/measurement/common/generic_query_measurement.go (hunk 3 above)
Comment (aleksandra-malinowska) [reply clusterloader2/pkg/measurement/common/generic_query_measurement.go:110]: Done

File: clusterloader2/pkg/measurement/common/generic_query_measurement.go [hunk 4]
```diff
@@ -40,26 +45,61 @@ func init() {
...
+		if err != nil {
+			return err
+		}
+		name, err := util.GetString(q, "name")
+		if err != nil {
+			return err
+		}
+		hasThreshold := false
+		threshold, err := util.GetFloat64(q, "threshold")
+		if err != nil {
```
Comment (mborsz) [inline clusterloader2/pkg/measurement/common/generic_query_measurement.go:91]: I think we should distinguish "not found" error (ignore) and other errors (fail). There is a function in util package to check if error type is not found.

**Comment by mborsz** [comment]
Looks good to me. Please fix one comment, fix linter and squash commits.

**Comment by mborsz** [comment]
LGMT from my side. I would like @jprzychodzen to also take a look.

File: clusterloader2/pkg/util/util.go [hunk 5]
```diff
@@ -71,7 +71,12 @@ func GetMap(dict map[string]interface{}, key string) (map[string]interface{}, er
 	return getMap(dict, key)
 }
 
-// GetStringArray tries to return value from map of type map with casting to a string using fmt.Sprintf. If value doesn't exist, error is returned.
+// GetMapArray tries to return value from map of type []map. If value doesn't exist, error is returned.
+func GetMapArray(dict map[string]interface{}, key string) ([]map[string]interface{}, error) {
+	return getMapArray(dict, key)
+}
+
+// GetStringArray tries to return value from map cast to a string using fmt.Sprintf. If value doesn't exist, error is returned.
```
Comment (jprzychodzen) [inline clusterloader2/pkg/util/util.go:79]: Are you sure you want to change this comment?

File: clusterloader2/pkg/util/util.go [hunk 6]
```diff
@@ -129,11 +134,33 @@ func getMap(dict map[string]interface{}, key string) (map[string]interface{}, er
 
 	mapValue, ok := value.(map[string]interface{})
 	if !ok {
-		return nil, fmt.Errorf("type assertion error: %v is not a string", value)
```
Comment (jprzychodzen) [inline clusterloader2/pkg/util/util.go:132]: Why this change?

File: clusterloader2/pkg/util/util.go (hunk 5 above)
Comment (aleksandra-malinowska) [reply clusterloader2/pkg/util/util.go:79]: I think I did, but not to this :) I meant cast to a string slice (as it doesn't return a map, nor accept a map of maps). Does that make sense?

File: clusterloader2/pkg/util/util.go (hunk 6 above)
Comment (jprzychodzen) [reply clusterloader2/pkg/util/util.go:132]: Ach, sorry, you are right. I've confused this part of a code with different one that was recently touched.

You are correct, we are casting to map[string]interface{}.

File: clusterloader2/pkg/util/util.go (hunk 5 above)
Comment (jprzychodzen) [reply clusterloader2/pkg/util/util.go:79]: Right, it should include a string slice information. Thanks.

**Comment by jprzychodzen** [comment]
Thanks for this change, single nit if you are improving the comment.

/lgtm
/hold
